src/sinks/nats.rs

```diff
@@ -293,5 +274,306 @@ mod integration_tests {
             ack_counter.load(std::sync::atomic::Ordering::Relaxed),
             num_events
         );
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn nats_no_auth() {
+        trace_init();
+
+        let subject = format!("test-{}", random_string(10));
+
+        let conf = NatsSinkConfig {
+            encoding: EncodingConfig::from(Encoding::Text),
+            connection_name: "".to_owned(),
+            subject: subject.clone(),
+            url: "nats://127.0.0.1:4222".to_owned(),
+            tls: None,
+            auth: None,
+        };
+
+        let r = publish_and_check(conf).await;
+        assert!(r.is_ok());
```

Comment:
Hmm, I see that you want some of the tests to explicitly fail, which is useful. However, this assertion (and below)will only report _that_ an error happened, not _what_ error happened. Maybe something like this? This will both report the error encountered and also include that it happened while calling `publish_and_check`.
```suggestion
        assert_eq!(publish_and_check(conf).await, Ok(()));
```

Reply:
The suggested snippet doesn't compile:
```
   Compiling vector v0.20.0 (/home/seeyarh/src/github.com/seeyarh/vector)
error[E0308]: mismatched types
   --> src/sources/nats.rs:263:9
    |
263 |         assert_eq!(publish_and_check(conf).await, Ok(()));
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `rkyv::result::ArchivedResult`, found enum `std::result::Result`
    |
    = note: expected enum `rkyv::result::ArchivedResult<_, _>`
               found enum `std::result::Result<(), _>`
    = note: this error originates in the macro `assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)
help: try using a variant of the expected enum
    |
39  |                 if !(*left_val == rkyv::result::ArchivedResult::Ok(*right_val)) {
    |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
39  |                 if !(*left_val == rkyv::result::ArchivedResult::Err(*right_val)) {
```

What about this for the passing tests
```
let r = publish_and_check(conf).await;
assert!(r.is_ok(), "publish_and_check failed, expected Ok(()), got: {:?}", r);
```

and this for the failing tests
```
        let r = publish_and_check(conf).await;
        assert!(
            matches!(r, Err(BuildError::Connection { .. })),
            "publish_and_check failed, expected BuildError::Connection, got: {:?}",
            r
        );
```

This reports a message like the following when the assert fails
```
---- sources::nats::integration_tests::nats_userpass_auth_valid stdout ----
thread 'sources::nats::integration_tests::nats_userpass_auth_valid' panicked at 'publish_and_check failed, expected Ok(()), got: Err(Connection { source: Os { code: 111, kind: ConnectionRefused, message: "Connection refused" } })', src/sources/nats.rs:264:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```